# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [228]:
def najdi_prvocisla(n,prvocisla = [2]):
    i = prvocisla[-1]
    while len(prvocisla) < n:
        i += 1
        for prvocislo in prvocisla:
            if i % prvocislo == 0:
                break
        else:
            prvocisla.append(i)
    return prvocisla[:n]

In [231]:
najdi_prvocisla(8)

[2, 3, 5, 7, 11, 13, 17, 19]


# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [160]:
%%writefile du1.py
def my_sqrt(a,n):
    x = a
    for i in range(n):
        x = ((a/x)+x)/2
    return x

def pi_from_polygon(n):
    a = 1
    b = 1
    v = my_sqrt(a**2 - (a / 2)**2, 100)
    s = 6 * b * v / 2
    #b = my_sqrt((a / 2)**2 + (1 - v)**2, 100)
    for i in range(n):
        b = my_sqrt((b / 2)**2 + (1 - v)**2, 100)
        #print(b)
        v = my_sqrt(a**2 - (b / 2)**2, 100)
        #print(v)
    s = 6 * 2**n * b * v / 2
    return s

def pi_from_series(n):
    
    sum = 0
    a = (1 / 2**4)

    for i in range(n):
        if i == 0:
            sum += (a / (2 * (i+1) + 1))
        else:
            a = a * ((2 * (i+1) - 3) / (2 * (i+1))) * (1 / 2**2)
            sum += (a / (2 * (i+1) + 1))
    
    return 12 * (- (my_sqrt(3,1000) / 8) + (1 / 2) - sum)



Writing du1.py


In [164]:
%time pi_from_polygon(10)

%timeit pi_from_polygon(10)

CPU times: user 1.76 ms, sys: 0 ns, total: 1.76 ms
Wall time: 1.79 ms
552 µs ± 8.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [166]:
%%time
pi_from_series(15)

CPU times: user 487 µs, sys: 0 ns, total: 487 µs
Wall time: 498 µs


3.1415926535900365

In [167]:
%%timeit
pi_from_series(15)

306 µs ± 5.41 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)



# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [169]:
import math
n = 0
while True:
    pi_approx = pi_from_series(n)
    x = math.floor(- math.log10(pi_approx - math.pi))
    print(f"n = {n}, pi = {pi_approx}, počet správných cifer = {x}")
    if x >= 8:
        break
    n += 1

n = 0, pi = 3.401923788646684, počet správných cifer = 0
n = 1, pi = 3.1519237886466844, počet správných cifer = 1
n = 2, pi = 3.1425487886466845, počet správných cifer = 3
n = 3, pi = 3.1417117350752557, počet správných cifer = 3
n = 4, pi = 3.1416100098148387, počet správných cifer = 4
n = 5, pi = 3.1415954446070975, počet správných cifer = 5
n = 6, pi = 3.1415931337808694, počet správných cifer = 6
n = 7, pi = 3.1415927403902137, počet správných cifer = 7
n = 8, pi = 3.1415926698836163, počet správných cifer = 7
n = 9, pi = 3.1415926567409396, počet správných cifer = 8



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.


In [189]:
def fibonacci(n):
    a = 0
    b = 1
    for i in range(n+1):
        yield a
        a, b = b, a+b

In [186]:
f = fibonacci(10)
print(next(f))
print(next(f))
print(next(f))
print(next(f))
print(next(f))
print(next(f))
print(next(f))


0
1
1
2
3
5
8



# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).


In [222]:
cisla = [i for i in range(11)]
print(cisla)

moc = [i**2 for i in range(11)]
print(moc)

prv = [i for i in range(101) if all(i % n != 0 for n in range(2,i-1))]
print(prv)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
[0, 1, 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]



# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

In [193]:
%%writefile neco.txt
hello world

Overwriting neco.txt


In [224]:
with open('neco.txt', 'r') as f:
    print(f.read())

with open('necojineho.txt', 'w') as g:
    g.write("ahoj\n")
    g.write(f"{cisla}\n{moc}\n{prv}\n")

hello world



# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [134]:
# kostra k doplnění, místa označená TODO doplňte

class Operator:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass
    
    def __add__(self, other):
        # TODO (použijte Addition)
        return Addition(self,other)
    
    def __sub__(self, other):
        # TODO (použijte Addition a Multiplication se zápornou konstantu)
        return Addition(self,Multiplication(other,-1))
    
    def __mul__(self, other):
        # TODO (použijte Multiplication)
        return Multiplication(self,other)
    
class BinaryOperator(Operator):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Operator):
    def __init__(self, argument):
        self.argument = argument
    
    def __str__(self):
        return type(self).__name__ + "(" + str(self.argument) + ")"
        
class Addition(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)+(pravy_argument)
        return str(self.levy_argument) + " + " + str(self.pravy_argument)
        
    def diff(self):
        # TODO
        return str(self.levy_argument.diff()) + " + " + str(self.pravy_argument.diff())
    
class Multiplication(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)*(pravy_argument)
        return "(" + str(self.levy_argument) + ")*(" + str(self.pravy_argument) + ")"

    def diff(self):
        # TODO
        return "(" + str(self.levy_argument.diff()) + ")*(" + str(self.pravy_argument) + ") + (" + str(self.levy_argument) + ")*(" + str(self.pravy_argument.diff()) + ")"
    
class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        # TODO vypíšeme něco jako value
        return str(self.value)
        
    def diff(self):
        # TODO
        return "0"
    
class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        # TODO vypíšeme něco jako x\
        return "x"   
        
    def diff(self):
        # TODO
        return "1"
    
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        # TODO vypíšeme něco jako argument^exponent
        return "(" + str(self.argument) + ")" + "^" + str(self.exponent)
    
    def diff(self):
        # TODO
        return "(" + str(self.exponent) + ")*" + str(mocnina(self.argument, self.exponent-1))
    
class sin(UnaryOperator):
    def diff(self):
        # TODO
        return "(" + str(cos(self.argument)) + ")*(" + str(self.argument.diff()) + ")"

class cos(UnaryOperator):
    def diff(self):
        # TODO
        return "((-1)*" + str(sin(self.argument)) + ")*(" + str(self.argument.diff()) + ")"

class exp(UnaryOperator):
    def diff(self):
        # TODO
        return "(" + str(exp(self.argument)) + ")*(" + str(self.argument.diff()) + ")"
    
class ln(UnaryOperator):
    def diff(self):
        # TODO
        return str(frc(1,self.argument))

class frc(UnaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako 1/(argument)
        return "1 / (" + self + ")"
    
    def diff(self):
        # TODO
        pass

In [135]:
x = identity()
f2 = mocnina(x,2)
print(f2)


(x)^2


In [136]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())

sin(x + cos((x)*(x))) + 1
(cos(x + cos((x)*(x))))*(1 + ((-1)*sin((x)*(x)))*((1)*(x) + (x)*(1))) + 0


můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [137]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


(x)^2 + (2)*(x) + 1 + (2)*((x)^2)
(2)*(x)^1 + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*((2)*(x)^1)


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [138]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())


(x)^2 + (2)*(x) + exp(x)
(2)*(x)^1 + (0)*(x) + (2)*(1) + (exp(x))*(1)


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...